src/types.rs

```diff
@@ -233,6 +233,12 @@ where
     ///
     fn key_use(&self) -> Option<&JU>;
 
+    ///
+    /// Returns the algorithm (e.g. ES512) this key must be used with, or `None` if no algorithm
+    /// constraint was specified.
+    ///
+    fn check_compatibility(&self, signing_algorithm: &JS) -> Result<(), &'static str>;
```

Comment:
this is a breaking change to a public trait, but I don't see a way to implement the desired functionality without adding a new method to the trait. to avoid a major version bump, let's put this behind a new optional feature flag called `jwk-alg` or something.

for simplicity and consistency with the rest of this trait, I think I'd prefer this function to be:
```rust
fn signing_alg(&self) -> Option<&JS>;
```
with that API, I think the rest of the logic can be folded into `filter_keys`

Reply:
> for simplicity and consistency with the rest of this trait, I think I'd prefer this function to be: fn signing_alg(&self) -> Option<&JS>;

That's what I did at first, but I believe there's an issue: with this API, we can't differentiate a missing alg field from an alg field with an encryption algorithm. I think it's an issue, as if the key does not have a key use, but has an encryption `alg`, we probably shouldn't use it for signing. I can think of multiple ways to address this issue:
 - return an `Option<Option<&JS>>`, where `None` is a missing field and `Some(None)` is an encryption algorithm
 - return an `Option<&KeyAlgorithm<JS>>`, with `enum KeyAlgorithm<JS> { None, Signature(JS), Encryption }`
 - return an `Option<&KeyAlgorithm<JS, JE>>`, with `enum KeyAlgorithm<JS, JE> { None, Signature(JS), Encryption(JE) }`. this is my favorite one, but it requires adding one more generic parameter to the JsonWebKey trait, which would require a major version bump
 - let the trait implementer decide (which is what I did)

Reply:
ah yes, that makes sense (`None` being ambiguous). what if we define a simpler enum for this function to return:
```rust
pub enum JsonWebKeySigningAlgorithm<JS, JT>
where
    JS: JwsSigningAlgorithm<JT>,
    JT: JsonWebKeyType,
{
  // sigh, PhantomData seems unavoidable here, as it would be in KeyAlgorithm
  Algorithm(JS, PhantomData<JT>),
  Unspecified,
  Unsupported,
}
```
the `Unspecified` variant instructs this library to allow the key to be used with any compatible signature algorithm (assuming the key use also allows signatures), while the `Unsupported` variant means the associated algorithm doesn't support signing (i.e., because it's a JWE algorithm). this will accomplish a similar purpose as `KeyAlgorithm` but without introducing a new type parameter. I agree that if we didn't have to worry about backward compatibility, `KeyAlgorithm` feels like the most natural approach (and may be worth adopting in a future major version).

Reply:
Yep, that was option 2 on my list, I'm happy to make that change

Reply:
sounds good!

Reply:
> that was option 2 on my list

I don't think it should return an option though since I'm not sure what `None` would mean in that context

Reply:
indeed, my bad